### PR TITLE
Refactors SM gas behavior to be datum based instead of list based + powerloss co2 buff

### PR DESCRIPTION
**Pull request by vincentiusvin**
## About The Pull Request
Title!
The CO2 thing is there because it makes my job much easier. Can probably find a way to make it move slowly if a maint insist on it. Prefer not to though.

~~Drafting because I want to make a second PR that have more sweeping changes (clean vars up, make a simpler formula for damage and heat production, delete underused behaviors, etc). Would honestly prefer if both this and that gets merged at the same time but I'm separating it out since it might be rejected. Or maybe ill combine it here we'll see.~~
Ignore that, looks like i can keep this one absolutely atomic.

## Why It's Good For The Game
Had a lot of trouble when trying to document the SM gas interactions into the wiki, the interactions are all scattered and tracking down everything a gas does is extremely annoying. Hopefully this fixes that.

## Changelog
:cl:
balance: CO2 powerloss inhibition now works immediately based on gas composition instead of slowly ramping up.
refactor: refactored how the SM fetches it's gas info and data. No changes expected except for the co2 thing.
/:cl:

### CODE REVIEW

**Comment by Kapu1178** [comment]
I don't know exactly how hard the SM hits the profiler but it might be worth compressing all of the datum info down into a static list to reduce datum var accesses.

**Comment by vincentiusvin** [comment]
SM process is called relatively rarely we'll be fine without micro ops I think.

File: code/modules/power/supermatter/supermatter_process.dm [hunk 1]
```diff
@@ -53,14 +54,38 @@
...
+
+		power_transmission_bonus = 0
+		dynamic_heat_modifier = 0
+		dynamic_heat_resistance = 0
+		gasmix_power_ratio = 0
+		powerloss_dynamic_scaling = 0
+		
+		for (var/gas_path in absorbed_gasmix.gases)
+			var/datum/sm_gas/sm_gas = GLOB.sm_gas_behavior[gas_path]
+			gas_comp[gas_path] = absorbed_gasmix.gases[gas_path][MOLES] / combined_gas
```
Comment (Zirok-BYOND) [inline code/modules/power/supermatter/supermatter_process.dm:68]: I suggest another name for this variable: portion_of_total would work as well. gas_fraction; gas_share; gas_percentage; portion_of_total

Comment (vincentiusvin) [reply code/modules/power/supermatter/supermatter_process.dm:68]: gas percentage sounds nice. Ill do it.

File: code/modules/power/supermatter/supermatter_gas.dm [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+
+/datum/sm_gas/bz
+	gas_path = /datum/gas/bz
+	heat_penalty = 5
+	transmit_modifier = -2
+	powermix = 1
+
+/// Start to emit radballs at a maximum of 30% chance per tick
+/datum/sm_gas/bz/extra_effects(obj/machinery/power/supermatter_crystal/sm, datum/gas_mixture/env)
+	if(sm.gas_comp[/datum/gas/bz] <= 0.4 && prob(30 * sm.gas_comp[/datum/gas/bz]))
```
[the remark this answers is not in the document]
Comment (vincentiusvin) [reply code/modules/power/supermatter/supermatter_gas.dm]: Oh shit right thanks.

**Pull request merged**